From 1e1e40184668d73ec49995fc6058e49fc9aef338 Mon Sep 17 00:00:00 2001
From: Russell King <rmk+kernel@arm.linux.org.uk>
Date: Tue, 14 Oct 2014 12:26:41 +0100
Subject: [PATCH 020/178] staging: etnaviv: provide etnaviv_gem_new_private()

etnaviv_gem_new_private() creates a private non-shmem gem object which
can be used to create these kinds of objects.  Fix up msm_gem_import()
to use it.

Signed-off-by: Russell King <rmk+kernel@arm.linux.org.uk>
---
 drivers/staging/etnaviv/etnaviv_gem.c | 32 ++++++++++++++++++++++----------
 drivers/staging/etnaviv/etnaviv_gem.h |  3 +++
 2 files changed, 25 insertions(+), 10 deletions(-)

diff --git a/drivers/staging/etnaviv/etnaviv_gem.c b/drivers/staging/etnaviv/etnaviv_gem.c
index b293793..083a0a6 100644
--- a/drivers/staging/etnaviv/etnaviv_gem.c
+++ b/drivers/staging/etnaviv/etnaviv_gem.c
@@ -687,6 +687,23 @@ fail:
 	return ERR_PTR(ret);
 }
 
+int etnaviv_gem_new_private(struct drm_device *dev, size_t size, uint32_t flags,
+	struct etnaviv_gem_object **res)
+{
+	struct drm_gem_object *obj;
+	int ret;
+
+	ret = etnaviv_gem_new_impl(dev, size, flags, &obj);
+	if (ret)
+		return ret;
+
+	drm_gem_private_object_init(dev, obj, size);
+
+	*res = to_etnaviv_bo(obj);
+
+	return 0;
+}
+
 static void etnaviv_gem_prime_release(struct etnaviv_gem_object *etnaviv_obj)
 {
 	if (etnaviv_obj->vaddr)
@@ -710,19 +727,15 @@ struct drm_gem_object *msm_gem_import(struct drm_device *dev,
 		struct dma_buf_attachment *attach, struct sg_table *sgt)
 {
 	struct etnaviv_gem_object *etnaviv_obj;
-	struct drm_gem_object *obj;
 	int ret, npages;
 	size_t size = PAGE_ALIGN(attach->dmabuf->size);
 
-	ret = etnaviv_gem_new_impl(dev, size, ETNA_BO_WC, &obj);
-	if (ret)
-		goto fail;
-
-	drm_gem_private_object_init(dev, obj, size);
+	ret = etnaviv_gem_new_private(dev, size, ETNA_BO_WC, &etnaviv_obj);
+	if (ret < 0)
+		return ERR_PTR(ret);
 
 	npages = size / PAGE_SIZE;
 
-	etnaviv_obj = to_etnaviv_bo(obj);
 	etnaviv_obj->ops = &etnaviv_gem_prime_ops;
 	etnaviv_obj->sgt = sgt;
 	etnaviv_obj->pages = drm_malloc_ab(npages, sizeof(struct page *));
@@ -736,11 +749,10 @@ struct drm_gem_object *msm_gem_import(struct drm_device *dev,
 	if (ret)
 		goto fail;
 
-	return obj;
+	return &etnaviv_obj->base;
 
 fail:
-	if (obj)
-		drm_gem_object_unreference_unlocked(obj);
+	drm_gem_object_unreference_unlocked(&etnaviv_obj->base);
 
 	return ERR_PTR(ret);
 }
diff --git a/drivers/staging/etnaviv/etnaviv_gem.h b/drivers/staging/etnaviv/etnaviv_gem.h
index f454881..367d4ab 100644
--- a/drivers/staging/etnaviv/etnaviv_gem.h
+++ b/drivers/staging/etnaviv/etnaviv_gem.h
@@ -106,4 +106,7 @@ struct etnaviv_gem_submit {
 	} bos[0];
 };
 
+int etnaviv_gem_new_private(struct drm_device *dev, size_t size, uint32_t flags,
+	struct etnaviv_gem_object **res);
+
 #endif /* __ETNAVIV_GEM_H__ */
-- 
2.6.2

